Add bulk XCom deletion to Execution API

justinpakzad opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

This PR implements bulk XCom deletion functionality. The issue was that it's not possible to delete all XComs for a dag run using the `XCom.delete()` method since task_id and key are required arguments. Given there is no direct database access, adding bulk deletion functionality would allow users to run clean up tasks at the end of their Dags, which can help prevent db bloat when the Xcoms don't really have much archival value. 

I added a new `bulk_delete_xcoms()` endpoint to the execution API with task_id and key as optional query parameters. This allows a user to just pass in dag_id and run_id to delete all XComs associated with that run.

A `delete_all()` method was added to the BaseXCom class as well as the associated method in the client which calls the execution API. New message type was added to the supervisor and trigger job runner.

I created tests to cover the different deletion patterns: deleting all XComs for a run, deleting all keys for a specific task, and deleting a specific key across all tasks. Cadwyn migration file was also updated with the new change.

closes:  #57812
<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

justinpakzad committed:
Implemented bulk deletion for xcoms

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -424,14 +424,27 @@ def delete_xcom(
+    """
+    Delete XCom entry(ies).
+
+    This endpoint allows specifying `~` as the task_id or key to enable bulk deletion.
```
I do not think this is intuitive at all. It is possible to have a key / value as "~", which makes it confusing as well. Not passing any parameters to `XCom.delete()` would likely be a better way to do this.

justinpakzad replied:
Got it. I updated the endpoint to take those args as optional query params and removed the '~' logic completely. The only issue I ran into was that the current `has_xcom_access()` dependency (which is not currently implemented) is applied to the router and wouldn't be compatible with the new version of delete endpoint. I moved it to individual endpoints for now and excluded it from delete. Not sure if that's the best approach but given it's not even implemented I thought it made sense. Curious to hear your thoughts.

justinpakzad committed:
Removed wild card  logic from task execution api delete endpoint and added optional query params

justinpakzad committed:
added test for Xcom delete with a specified map index

justinpakzad committed:
prek formatting

justinpakzad committed:
fixed mypy error with params dict

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -66,7 +66,6 @@ async def has_xcom_access(
         status.HTTP_403_FORBIDDEN: {"description": "Task does not have access to the XCom"},
         status.HTTP_404_NOT_FOUND: {"description": "XCom not found"},
     },
-    dependencies=[Depends(has_xcom_access)],
```
Why has this been changed?

justinpakzad replied:
Because the `has_xcom_access` function expects xcom_key to be a path parameter: `xcom_key: Annotated[str, Path(alias="key", min_length=1)]` and `task_id` to have a value. This wouldn't work for bulk deleting xcoms (even if it was a separate endpoint) because it's applied at the router level. We could just change them to both be optional, but then that would apply to all the other endpoints even though they are actually required for those endpoints. We might need a separate dependency to handle the has access for bulk deletion or find a way to conform the current one so it works for all. Feel free to correct me if I'm mistaken or you know of a better way.

uranusjr replied:
`has_xcom_access` does not currently do anything so it should work for `bulk_delete_xcoms` now if we change the function signature. I think eventually the bulk endpoint should have an access check anyway. Can you try if add `has_xcom_access` as a dependency work on the router if some arguments are made optional?

justinpakzad replied:
Yea, so it works as a dependency on the router when I change the function to be like:
```
async def has_xcom_access(
    dag_id: str,
    run_id: str,
    request: Request,
    task_id: str | None = None,
    key: str | None = None,
    token=JWTBearerDep,
) -> bool:
```

Basically need to remove `Annotated[str, Path(alias="key", min_length=1)]` since Path params are always required and then `task_id` & `key` (changed the name to key so it's recognized) need to be made optional.

uranusjr replied:
Let’s do this change and restore the previous `dependencies` declaration on the router then. That makes things easier in the future.

justinpakzad replied:
Sounds good. Will push those changes in a bit. Appreciate the help!

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -413,26 +417,40 @@ def set_xcom(
+    Delete XCom entry(ies).
+
+    Supports bulk deletion when task_id and/or key are not provided.
+    """
```
This is dangerous. This makes: `DELETE /example_dag/manual__2025-01-01T00:00:00` a valid call and can delete all xcoms if not done carefully.

Another issue I see is:
`DELETE /dag/run?task_id=task_a&key=return_value` earlier can now be `DELETE /dag/run?task_id=task_a` unintentionally, which will delete all xcoms. 

I think having another endpoint is the way to go here, something like a:
```python
@router.delete(
    "/{dag_id}/{run_id}/xcoms",
    description="Bulk delete XCom values,
)
def bulk_delete_xcoms(
    session: SessionDep,
    dag_id: str,
    run_id: str,
    task_id: Annotated[str | None, Query()] = None,
    key: Annotated[str | None, Query()] = None,
    map_index: Annotated[int | None, Query()] = None,
):
```

Does this not seem safer?

justinpakzad replied:
Got it, that makes sense. Maybe I was being a bit naive and didn't consider some of the potential dangers here. Happy to create the separate endpoint. The only thing I'm still uncertain about is, if we create a new endpoint, but still just use `XComOperations.delete()` and determine which endpoint to use (bulk or single) based on the parameters passed, does that actually address the safety concern you have? I think we might need to have a `XComOperations.bulk_delete()` method that explicitly calls the bulk endpoint, thus preventing  any accidental bulk operations. Curious to hear what you think.

amoghrajesh replied:
You can have a `delete_all` in there, in any case the API client is internal module as is wired in to be called from task sdk supervisor, so it won't really be exposed.

justinpakzad replied:
Sounds good, I made the changes.

justinpakzad replied:
Lmk when you have a chance to check the changes. Cheers.

justinpakzad committed:
updated xcom bulk delete tests

justinpakzad committed:
Added separate bulk delete endpoint and modified relevant code in the task-sdk

justinpakzad committed:
Fixed test_message_types_in_dag_processor test, prek and resolved conflict

justinpakzad committed:
Added missing jwt token to bulk endpoint

justinpakzad committed:
Fixed merge conflicts

justinpakzad committed:
Updated the has_xcom_access func to make task_id and key optional

amoghrajesh commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -473,6 +475,8 @@ def _handle_request(self, msg: ToTriggerSupervisor, log: FilteringBoundLogger, r
             self.client.variables.set(msg.key, msg.value, msg.description)
         elif isinstance(msg, DeleteXCom):
             self.client.xcoms.delete(msg.dag_id, msg.run_id, msg.task_id, msg.key, msg.map_index)
+        elif isinstance(msg, BulkDeleteXCom):
+            self.client.xcoms.delete_all(msg.dag_id, msg.run_id, msg.task_id, msg.key, msg.map_index)
         elif isinstance(msg, GetXCom):
```
Is there a reason not to support this in dag processor too if we do it in triggerer?

justinpakzad replied:
I didn't see the existing DeleteXCom supported in the processor so I didn't include it. Happy to add it in if it makes sense to.

justinpakzad replied:
Any thoughts on this?

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/execution_api/versions/v2026_04_06.py:
```diff
@@ -95,3 +95,13 @@ def ensure_start_date_in_dag_run(response: ResponseInfo) -> None:  # type: ignor
+    description = __doc__
+
+    instructions_to_migrate_to_previous_version = (
+        endpoint("xcoms/{dag_id}/{run_id}", ["DELETE"]).didnt_exist,
```
Looks ok

amoghrajesh commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -437,3 +437,37 @@ def delete_xcom(
+
+@router.delete(
+    "/{dag_id}/{run_id}",
+    responses={status.HTTP_404_NOT_FOUND: {"description": "XComs not found"}},
```
We should be raising this somewhere right?

amoghrajesh replied:
Or remove it from here if never raised

justinpakzad replied:
Yea, I think removing it makes sense. I originally included it to match the single delete endpoint (which doesn't raise either).

justinpakzad replied:
Removed the responses.

amoghrajesh reviewed (commented):
Looks better but I have some more qns.

justinpakzad committed:
Removed response that never gets raised

amoghrajesh commented on task-sdk/src/airflow/sdk/api/client.py:
```diff
@@ -568,6 +568,30 @@ def delete(
+        task_id: str | None = None,
+        key: str | None = None,
+        map_index: int | None = None,
+    ):
```
```suggestion
    ) -> OKResponse :
```

justinpakzad replied:
Fixed with the new XComDeleteCountResponse, which addresses your comment below.

amoghrajesh commented on task-sdk/src/airflow/sdk/api/client.py:
```diff
@@ -568,6 +568,30 @@ def delete(
+        # Any error from the server will anyway be propagated down to the supervisor,
+        # so we choose to send a generic response to the supervisor over the server response to
+        # decouple from the server response string
+        return OKResponse(ok=True)
```
Wonder if it will be beneficial to return number of deleted rows instead?

justinpakzad replied:
Yea, good call. I made that update. The endpoint now returns an `XComDeleteCountResponse` with the number of deleted rows. Updated everything else accordingly.

amoghrajesh commented on task-sdk/tests/task_sdk/execution_time/test_supervisor.py:
```diff
@@ -1732,6 +1733,27 @@ class RequestTestCase:
         ),
         test_id="delete_xcom",
     ),
+    RequestTestCase(
+        message=BulkDeleteXCom(
+            dag_id="test_dag",
+            run_id="test_run",
+        ),
+        client_mock=ClientMock(
+            method_path="xcoms.delete_all",
+            args=("test_dag", "test_run", None, None, None),
+            response=OKResponse(ok=True),
+        ),
+        test_id="bulk_delete_xcoms",
+    ),
+    RequestTestCase(
+        message=BulkDeleteXCom(dag_id="test_dag", run_id="test_run", task_id="t1"),
+        client_mock=ClientMock(
+            method_path="xcoms.delete_all",
+            args=("test_dag", "test_run", "t1", None, None),
+            response=OKResponse(ok=True),
+        ),
+        test_id="bulk_delete_xcoms",
+    ),
```
Both have same test_id

justinpakzad replied:
Good find. Fixed.

amoghrajesh commented on task-sdk/tests/task_sdk/execution_time/test_supervisor.py:
```diff
@@ -1732,6 +1733,27 @@ class RequestTestCase:
         ),
         test_id="delete_xcom",
     ),
+    RequestTestCase(
+        message=BulkDeleteXCom(
+            dag_id="test_dag",
+            run_id="test_run",
+        ),
+        client_mock=ClientMock(
+            method_path="xcoms.delete_all",
+            args=("test_dag", "test_run", None, None, None),
+            response=OKResponse(ok=True),
+        ),
+        test_id="bulk_delete_xcoms",
+    ),
+    RequestTestCase(
+        message=BulkDeleteXCom(dag_id="test_dag", run_id="test_run", task_id="t1"),
+        client_mock=ClientMock(
+            method_path="xcoms.delete_all",
+            args=("test_dag", "test_run", "t1", None, None),
+            response=OKResponse(ok=True),
+        ),
+        test_id="bulk_delete_xcoms",
+    ),
```
Can you also add in a test for `map_index` variation?

justinpakzad replied:
Added.

amoghrajesh reviewed (commented):
Apologies for the delay, last few comments, looking much better now

justinpakzad committed:
applied suggestions from review

justinpakzad committed:
Add bulk XCom deletion to Execution API
Add a DELETE /xcoms/{dag_id}/{run_id} endpoint that
deletes all XCom entries for a Dag run, with optional
task_id, key, and map_index filters. Returns the count
of deleted rows.

justinpakzad committed:
Add bulk XCom deletion to Execution API
Add a DELETE /xcoms/{dag_id}/{run_id} endpoint that
deletes all XCom entries for a Dag run, with optional
task_id, key, and map_index filters. Returns the count
of deleted rows.

kaxil commented on task-sdk/src/airflow/sdk/api/client.py:
```diff
@@ -652,6 +653,27 @@ def delete(
+        """Bulk delete XCom values via the API server."""
+        params: dict[str, str | int] = {}
+
+        if map_index is not None and map_index >= 0:
```
`delete_all` drops `map_index` when it's `-1` (the `>= 0` guard), but the bulk endpoint treats a missing `map_index` as "all map indexes" and only filters when a value is sent. So `delete_all(dag_id, run_id, map_index=-1)`, meant to target non-mapped XComs, silently deletes every map index for the run instead. The `>= 0` guard fits `delete()` where the server defaults `map_index` to `-1`, but here `None` is the "unspecified" sentinel, so I think this should be `if map_index is not None:`.

justinpakzad replied:
Good call, updated. I also added a test to the `test_client.py` file as I noticed client-side coverage was missing.

justinpakzad committed:
Remove map_index filter and add test for client

justinpakzad committed:
Remove map_index filter and add test for client

kaxil commented on task-sdk/src/airflow/sdk/bases/xcom.py:
```diff
@@ -572,3 +574,42 @@ def delete(
+        """
+        from airflow.sdk.execution_time.task_runner import SUPERVISOR_COMMS
+
+        msg = SUPERVISOR_COMMS.send(
```
`delete` calls `cls.purge(xcom_result)` before sending the message, which is how a custom backend drops the payload it wrote outside the database (`XComObjectStorageBackend.purge` unlinks the object storage file). `delete_all` skips that, so on a custom backend the rows go away and the objects behind them are orphaned with nothing left pointing at them, which is the population this feature is aimed at. Can it purge the matched entries, or at minimum carry the note `XComModel.clear` has: "This **will not** purge any data from a custom XCom backend"? The docs sentence that would need amending is xcoms.rst:113, "This will be called as part of ``delete``".

justinpakzad replied:
I've updated the docstring and xcoms.rst:113 to indicate that this will not purge any data for custom Xcom backends. Bulk purging can be implemented in a follow up as I felt it would add another layer of complexity that would be better suited for a separate PR.

kaxil commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -392,6 +394,7 @@ def from_api_response(cls, response: HITLDetailResponse) -> HITLDetailResponseRe
     | GetVariableKeys
     | PutVariable
     | DeleteXCom
+    | BulkDeleteXCom
```
Adding `BulkDeleteXCom` here makes it sendable from a trigger, but `XComDeleteCountResult` never went into `ToTriggerRunner` above and the test parks it in `in_task_but_not_in_trigger_runner` instead. Since `handle_bulk_delete_xcom` returns a response (unlike `handle_delete_xcom`, which returns `None`, which is why `DeleteXCom` needs no result type) and the runner validates replies with `TypeAdapter(ToTriggerRunner)`, a trigger calling `delete_all` gets `Input tag 'XComDeleteCountResult' does not match any of the expected tags` after the rows are already gone. The one comparable pair is `GetPreviousTI`/`PreviousTIResult`, but `get_previous_ti` only exists on `RuntimeTaskInstance`, so it has no trigger-side caller. `delete_all` is a classmethod callable from anywhere, and `TriggerRunner.init_comms` sets `task_runner.SUPERVISOR_COMMS = self.comms_decoder`, so the call does go through from a trigger. Adding `| XComDeleteCountResult` to `ToTriggerRunner` and dropping the test exclusion fixes it, or the triggerer wiring could come back out.

justinpakzad replied:
Added it to the `ToTriggerRunner` and dropped it from the test exclusion.

kaxil commented on task-sdk/src/airflow/sdk/bases/xcom.py:
```diff
@@ -572,3 +574,42 @@ def delete(
+            will be deleted. Pass *None* (default) to delete all keys.
+        :param map_index: Optional map index filter. If provided, only XComs with this
+            map index will be deleted. Pass *None* (default) to delete all map indexes.
+        :return: Number of XCom entries deleted.
```
Worth a caveat in here: with no `task_id` or `key` filter this also removes the `dag_result=True` rows that `GET /dags/{dag_id}/dagRuns/{dag_run_id}/wait` serves. `_serialize_response` guards with `if result_xcoms := await self._serialize_xcoms():`, so a client waiting on the run gets `{"state": "success"}` with the `results` key absent rather than an empty dict, and the end-of-dag cleanup task described in the PR body is exactly the shape that triggers it. Skipping `dag_result` rows unless they are asked for, or at least documenting it and steering people to the filters, would avoid a silent hole that is hard to trace back here.

justinpakzad replied:
Good catch. I've added a new parameter `include_dag_result` which defaults to `False`, skipping `dag_result` rows unless asked for.

kaxil commented on task-sdk/src/airflow/sdk/api/client.py:
```diff
@@ -652,6 +653,27 @@ def delete(
+            params["key"] = key
+
+        resp = self.client.delete(url=f"xcoms/{dag_id}/{run_id}", params=params)
+        return XComDeleteCountResult(count=resp.json()["count"])
```
`Client.request` retries on `httpx.RequestError` and 5xx, and the server commits inside `create_session` before the response is written, so a DELETE whose response is lost gets replayed and the retry finds nothing left to delete. The caller then sees `count=0` after N rows actually went away, which `delete` above avoids only because it returns a fixed `OKResponse`. Worth documenting the count as best effort rather than exact, given the alternative is an idempotency key?

justinpakzad replied:
I've updated the docstring to state that the returned count is best effort and retried request can report fewer rows than were actually deleted.

kaxil commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -482,3 +482,36 @@ def delete_xcom(
+        query = query.where(XComModel.map_index == map_index)
+
+    result = session.execute(query)
+    count = getattr(result, "rowcount", 0)
```
Two things about producing and returning this count. `session.execute()` on a DELETE returns a `CursorResult`, which always carries `rowcount`, so the `getattr` fallback cannot fire, and if it ever could then reporting 0 would tell the caller nothing was deleted while rows were. Returning a bare dict also keeps the shape out of the OpenAPI spec, so codegen emits no model and the client has to reach for `resp.json()["count"]`; the sibling count routes in this API (`get_dr_count`, `get_task_instance_count`) just annotate `-> int` and let the client wrap it, as `TICount(count=resp.json())` does.

justinpakzad replied:
Removed the `getattr`, route returns an `int` directly. Added a cast because session.execute() returns Result[Any], which has no rowcount.

kaxil commented on airflow-core/src/airflow/api_fastapi/execution_api/versions/v2026_09_09.py:
```diff
@@ -0,0 +1,30 @@
+    description = __doc__
+
+    instructions_to_migrate_to_previous_version = (
+        endpoint("/xcoms/{dag_id}/{run_id}", ["DELETE"]).didnt_exist,
```
Nothing asserts the 404 contract this version file promises. `versions/v2026_06_30/test_variables.py` has the pattern for the equivalent new endpoint change: pin `Airflow-API-Version` to the previous release and assert 404. The versioning guide asks for tests covering both the new and the previous version, and this is the only thing that pins what older workers see.

justinpakzad replied:
I've added a test following the same pattern as `versions/v2026_06_30/test_variables.py`

kaxil commented on airflow-core/tests/unit/api_fastapi/execution_api/versions/head/test_xcoms.py:
```diff
@@ -617,6 +625,111 @@ def test_xcom_delete_endpoint(self, client, create_task_instance, session):
+        assert set(remaining) == {1}
+
 
 class TestXComTeamAccess:
```
Two gaps in the new tests. Both bulk cases create a single dag run, so removing `XComModel.run_id == run_id` from the route's WHERE clause would leave every assertion green; `test_xcom_delete_endpoint` just above does the right thing by creating a second dag and asserting its XCom survives. And `TestXComTeamAccess._url` below only builds the single XCom path, so the new write route never passes through `has_xcom_access` in any test even though this PR changes that function's signature.

justinpakzad replied:
I have added an another test that creates a second dag run and asserts it's XCom survives. I also added two additional tests to `TestXComTeamAccess` for the bulk path.

kaxil commented on airflow-core/src/airflow/api_fastapi/execution_api/routes/xcoms.py:
```diff
@@ -482,3 +482,36 @@ def delete_xcom(
+    map_index: Annotated[int | None, Query()] = None,
+):
+    """Bulk delete Xcom values."""
+    query = delete(XComModel).where(
```
`XComModel.clear` uses the same denormalized predicate so the shape is established, but the delta here is that without `task_id` the seek degrades to the `dag_id` prefix of `idx_xcom_task_instance(dag_id, task_id, run_id, map_index)`, so this walks the dag's whole XCom history to delete one run's worth and holds the locks for that long. `XComModel.set` already resolves the run first with `select(DagRun.id).where(DagRun.dag_id == ..., DagRun.run_id == ...)`, and `dag_run_id` is the leading primary key column, so filtering on that instead would make it an exact range. Once per dag run is not a hot path, but the deployments this feature targets are the ones with the biggest xcom tables.

justinpakzad replied:
Good find. I've updated it to first resolve the run and then to filter on `dag_run_id` as suggested.

justinpakzad committed:
Apply feedback

justinpakzad committed:
Apply feedback
